src/background.js

```diff
@@ -0,0 +1,142 @@
+    // Thus iterating from the last one (the most recent) is usually
+    // better, because players usually want to analyse their most recent
+    // games. So you will make less fetches
+    for (let i = archives.length - 1; i >= 0; i--) {
+      const games = (await fetchJSON(archives[i])).games;
+      for (let j = games.length - 1; j >= 0; j--) {
+        if (games[j].url === gameUrl) {
+          return games[j].pgn;
+        }
+      }
+    }
+  } catch (error) {
+    sendLogMessage(error);
+    return null;
+  }
+
+  return null;
+};
+
+const sendLogMessage = async (message) => {
+  await browser.tabs.executeScript({
+    code: `console.log("[Chess.com analyse at lichess]: ${message}");`,
+  });
+};
+
+const clickElement = async (querySelector) => {
+  await browser.tabs.executeScript({
+    code: `document.querySelector("${querySelector}").click();`,
+  });
+};
+
+const waitForElement = async (querySelector) => {
+  const elementFoundScript = `document.querySelector("${querySelector}") !== null;`;
+  const retryDelay = 100;
+
+  let timeLeft = 5000;
+  let [isElement] = await browser.tabs.executeScript({
+    code: elementFoundScript,
+  });
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise((r) => setTimeout(r, retryDelay));
+    [isElement] = await browser.tabs.executeScript({
+      code: elementFoundScript,
+    });
+    timeLeft -= retryDelay;
+  }
+
+  if (!isElement) sendLogMessage(`Cannot find \`${querySelector}\` element!`);
+
+  return isElement;
+};
+
+const waitAndClick = async (querySelector) => {
+  if (await waitForElement(querySelector)) await clickElement(querySelector);
+};
+
+let analysingState = false;
+
+const analyseGame = async (tab) => {
+  // After clicking on pageAction twice, second call won't be executed
+  if (analysingState) return;
+
+  analysingState = true;
+
+  await setLoadingState(true);
+
+  const [playerName] = await browser.tabs.executeScript({
+    code: `document.querySelector('[data-test-element="user-tagline-username"]').textContent;`,
+  });
+  const gameId = tab.url.split("?")[0];
+
+  const pgn = await getPGN(playerName, gameId);
+
+  await setLoadingState(false);
+
+  if (!pgn) {
+    sendLogMessage(`Game with id ${gameId} not found!`);
+    analysingState = false;
+    return;
+  }
+
+  await browser.tabs.create({ url: "https://lichess.org/paste" });
+
+  if (await waitForElement("[name='analyse']")) {
+    const [loggedIn] = await browser.tabs.executeScript({
+      code: `!document.querySelector("[name='analyse']").disabled;`,
+    });
+
+    if (loggedIn) await waitAndClick("[name='analyse']");
+
+    if (await waitForElement("[name='pgn']")) {
+      await browser.tabs.executeScript({
+        code: `document.querySelector("[name='pgn']").value = \`${pgn}\`;`,
+      });
+
+      await waitAndClick(".submit");
+
+      await waitAndClick("[for='analyse-toggle-ceval']");
+    }
+  }
+
+  analysingState = false;
```

Comment:
Is there a way that any of the previous operations fail and the code never gets to this line? This could cause the extension to be unusable after the user clicks the extension once and it fails at some step :smile: 

You could put this line in a `finally` block from `try...catch...finally`, which would make it always execute :smile:

Reply:
Resolved https://github.com/zeraye/chess-analysis-firefox-extension/pull/2/commits/c66c575bc572cdc520b3b59b3e2f568f6c68bf9a